### PR TITLE
Add Prometheus Exporter: Step 1

### PR DESCRIPTION
**Pull request by CunjunWang**
This PR implements exporter utils helper functions for the Prometheus Exporter project. The C++ file `prometheus_exporter_utils.cc` implements all the helper functions that will be called when we translate a OpenTelemetry (OTLP) metric to a Prometheus metric.

### Changes
1. Modified the `WORKSPACE` file and added C++ Prometheus Client as a dependency to the project. 
2. Modified `CMakeLists.txt` file in exporters folder to add Prometheus Exporter as a sub-project.
3. Added BUILD file in `exporters/prometheus` to support building project and the tests with Bazel.
4. Added `CMakeLists.txt` file in `exporters/prometheus` to support building project with CMake.
5. Added function header definition in `exporters/prometheus/include`
6. Implemented the header in `exporters/prometheus/src/prometheus_exporter_utils.cc`
7. Tested the implementation with unit tests in `exporters/prometheus/test/prometheus_exporter_utils_test.cc`
8. Added `CMakeLists.txt` file in `exporters/prometheus/test` to support building tests with CMake.
9. Modified `ci/setup_cmake.sh` to configure Prometheus C++ client as a dependency
10. Modified `.github/workflows/ci.yml` to execute Prometheus Exporter test sub-job
11. Modified `ci/do_ci.sh` to add Prometheus Exporter test setup scripts, and ignore Prometheus tests in `bazel.noexcept`

### CODE REVIEW

**Comment by CunjunWang** [comment]
For the CI, we actually need to have prometheus-cpp library installed in the CI environment because it's a project dependency

**Comment by reyang** [comment]
> For the CI, we actually need to have prometheus-cpp library installed in the CI environment because it's a project dependency

Could you add a GitHub Actions to cover that? You can potentially leverage a linux container and install the dependency you need for testing. Here goes and example of how to do that https://github.com/open-telemetry/opentelemetry-dotnet/tree/master/test/OpenTelemetry.Instrumentation.StackExchangeRedis.Tests.

File: WORKSPACE [hunk 1]
```diff
@@ -88,3 +88,16 @@ http_archive(
         "https://github.com/nlohmann/json/releases/download/v3.6.1/include.zip",
     ],
 )
+
+# C++ Prometheus Client library.
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive", "http_file")
+
+http_archive(
+    name = "com_github_jupp0r_prometheus_cpp",
+    strip_prefix = "prometheus-cpp-master",
```
Comment (g-easy) [inline WORKSPACE:97]: Please pin to a version (and add sha256), for consistency with other deps.

File: exporters/prometheus/include/opentelemetry/exporters/prometheus/prometheus_exporter_utils.h [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+#pragma once
+
+#include <string>
+#include <vector>
+
+#include "opentelemetry/sdk/metrics/record.h"
+#include "prometheus/metric_family.h"
+
+namespace prometheus_client = ::prometheus;
```
Comment (g-easy) [inline exporters/prometheus/include/opentelemetry/exporters/prometheus/prometheus_exporter_utils.h:25]: Is there a way to limit the scope of these? I'm worried about having them in a header file.

File: exporters/prometheus/src/prometheus_exporter_utils.cc [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+{
+/**
+ * Helper function to convert OpenTelemetry metrics data collection
+ * to Prometheus metrics data collection
+ *
+ * @param records a collection of metrics in OpenTelemetry
+ * @return a collection of translated metrics that is acceptable by Prometheus
+ */
+std::vector<prometheus_client::MetricFamily> PrometheusExporterUtils::TranslateToPrometheus(
+    std::vector<metric_sdk::Record> &records)
```
Comment (g-easy) [inline exporters/prometheus/src/prometheus_exporter_utils.cc:39]: Can records be a constref?

File: exporters/prometheus/src/prometheus_exporter_utils.cc [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+  }
+
+  return output;
+}
+
+// ======================= private helper functions =========================
+/**
+ * Set value to metric family according to record
+ */
+void PrometheusExporterUtils::SetMetricFamily(metric_sdk::Record &record,
```
Comment (g-easy) [inline exporters/prometheus/src/prometheus_exporter_utils.cc:64]: Can record be a constref?

File: exporters/prometheus/src/prometheus_exporter_utils.cc [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+                                              prometheus_client::MetricFamily *metric_family)
+{
+
+  auto origin_name = record.GetName();
+  auto sanitized   = SanitizeNames(origin_name);
+  if (origin_name != sanitized)
+  {
+    std::cout << "Sanitized metric name \"" << origin_name << "\" to \"" << sanitized << "\""
+              << std::endl;
```
Comment (g-easy) [inline exporters/prometheus/src/prometheus_exporter_utils.cc:73]: Please use \n over endl.

File: exporters/prometheus/src/prometheus_exporter_utils.cc [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+}
+
+/**
+ * Set time and labels to metric data
+ */
+void PrometheusExporterUtils::SetMetricBasic(prometheus_client::ClientMetric &metric,
+                                             std::chrono::nanoseconds time,
+                                             const std::string &labels)
+{
+  metric.timestamp_ms = time.count() / 1000;
```
Comment (g-easy) [inline exporters/prometheus/src/prometheus_exporter_utils.cc:284]: milliseconds = nanoseconds / 1000?

File: exporters/prometheus/src/prometheus_exporter_utils.cc [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+  {
+    metric.label.resize(label_pairs.size());
+    for (int i = 0; i < label_pairs.size(); ++i)
+    {
+      auto origin_name = label_pairs[i].first;
+      auto sanitized   = SanitizeNames(origin_name);
+      if (origin_name != sanitized)
+      {
+        std::cout << "Sanitized label name \"" << origin_name << "\" to \"" << sanitized << "\""
+                  << std::endl;
```
Comment (g-easy) [inline exporters/prometheus/src/prometheus_exporter_utils.cc:297]: \n

File: WORKSPACE (hunk 1 above)
Comment (CunjunWang) [reply WORKSPACE:97]: Sure. Where can I get SHA256 of this dependency repo, or do I generate a random one with some rules?

File: exporters/prometheus/src/prometheus_exporter_utils.cc (hunk 4 above)
Comment (CunjunWang) [reply exporters/prometheus/src/prometheus_exporter_utils.cc:64]: If I make record a constref here, I need to change all getter functions in Record class const functions. Record class was implemented by previous interns, and also used in other classes.

File: exporters/prometheus/src/prometheus_exporter_utils.cc (hunk 6 above)
Comment (CunjunWang) [reply exporters/prometheus/src/prometheus_exporter_utils.cc]: fixed. thanks

File: exporters/prometheus/src/prometheus_exporter_utils.cc (hunk 5 above)
Comment (CunjunWang) [reply exporters/prometheus/src/prometheus_exporter_utils.cc]: fixed

File: exporters/prometheus/src/prometheus_exporter_utils.cc (hunk 7 above)
Comment (CunjunWang) [reply exporters/prometheus/src/prometheus_exporter_utils.cc]: fixed

File: exporters/prometheus/src/prometheus_exporter_utils.cc (hunk 3 above)
Comment (CunjunWang) [reply exporters/prometheus/src/prometheus_exporter_utils.cc]: changed to a constref

File: exporters/prometheus/include/opentelemetry/exporters/prometheus/prometheus_exporter_utils.h (hunk 2 above)
Comment (CunjunWang) [reply exporters/prometheus/include/opentelemetry/exporters/prometheus/prometheus_exporter_utils.h]: removed it from header, and put it in implementation files

File: WORKSPACE (hunk 1 above)
Comment (CunjunWang) [reply WORKSPACE]: pin to v0.9.0, and sha256 added

File: ci/do_ci.sh [hunk 8]
```diff
@@ -40,6 +40,33 @@ elif [[ "$1" == "cmake.exporter.otprotocol.test" ]]; then
...
   make test
   exit 0
+elif [[ "$1" == "cmake.exporter.prometheus.test" ]]; then
+#  export DEBIAN_FRONTEND=noninteractive
+#  apt-get update
+#  apt-get install sudo
+#  apt-get install zlib1g-dev
+#  apt-get -y install libcurl4-openssl-dev
+  cd third_party
+  git clone https://github.com/jupp0r/prometheus-cpp
```
Comment (reyang) [inline ci/do_ci.sh:50]: Consider extract this logic to a separate file under the Prometheus exporter folder.

File: ci/do_ci.sh [hunk 9]
```diff
@@ -83,8 +110,8 @@ elif [[ "$1" == "bazel.legacy.test" ]]; then
   bazel test $BAZEL_TEST_OPTIONS -- //... -//exporters/otlp/...
   exit 0
 elif [[ "$1" == "bazel.noexcept" ]]; then
-  bazel build --copt=-fno-exceptions $BAZEL_OPTIONS //...
-  bazel test --copt=-fno-exceptions $BAZEL_TEST_OPTIONS //...
+  bazel build --copt=-fno-exceptions $BAZEL_OPTIONS -- //... -//exporters/prometheus/...
```
Comment (reyang) [inline ci/do_ci.sh:113]: Why would we want to take exception here?

Comment (CunjunWang) [reply ci/do_ci.sh:113]: When we run the test, we found some error handling and exceptions from the Prometheus client, which made this test always fail. We discussed this problem with Johannes last week, and he suggested us to consider excluding the Prometheus component in the noexcept test.

File: ci/do_ci.sh (hunk 8 above)
Comment (CunjunWang) [reply ci/do_ci.sh:50]: Sure, correct me if I misunderstand something:
1. keep the `elif [[ "$1" == "cmake.exporter.prometheus.test" ]];` block here
2. move the scripts of installing the Prometheus client (from `cd third_party` to `sudo make install`) to a separate script in `exporters/prometheus`. For example, `exporters/prometheus/install_prometheus_client.sh`
3. run `install_prometheus_client.sh` here

File: ci/do_ci.sh (hunk 9 above)
Comment (reyang) [reply ci/do_ci.sh:113]: Please add a comment here.

File: exporters/prometheus/src/prometheus_exporter_utils.cc [hunk 10]
```diff
@@ -0,0 +1,467 @@
...
+ */
+void PrometheusExporterUtils::SetMetricFamily(metric_sdk::Record &record,
+                                              prometheus_client::MetricFamily *metric_family)
+{
+
+  auto origin_name = record.GetName();
+  auto sanitized   = SanitizeNames(origin_name);
+  if (origin_name != sanitized)
+  {
+    std::cout << "Sanitized metric name \"" << origin_name << "\" to \"" << sanitized << "\"\n";
```
Comment (reyang) [inline exporters/prometheus/src/prometheus_exporter_utils.cc:74]: Is this intended or just for development/debugging purpose?

File: exporters/prometheus/src/prometheus_exporter_utils.cc [hunk 11]
```diff
@@ -0,0 +1,467 @@
...
+  if (!label_pairs.empty())
+  {
+    metric.label.resize(label_pairs.size());
+    for (int i = 0; i < label_pairs.size(); ++i)
+    {
+      auto origin_name = label_pairs[i].first;
+      auto sanitized   = SanitizeNames(origin_name);
+      if (origin_name != sanitized)
+      {
+        std::cout << "Sanitized label name \"" << origin_name << "\" to \"" << sanitized << "\"\n";
```
Comment (reyang) [inline exporters/prometheus/src/prometheus_exporter_utils.cc:297]: Same here, need to understand if we intend to keep the cout lines.

File: exporters/prometheus/src/prometheus_exporter_utils.cc (hunk 10 above)
Comment (CunjunWang) [reply exporters/prometheus/src/prometheus_exporter_utils.cc:74]: We added this printing in order to tell the user we have changed their metrics name or label names. We know this is not in the specification, and we can remove this.

Comment (CunjunWang) [reply exporters/prometheus/src/prometheus_exporter_utils.cc]: removed.

File: exporters/prometheus/src/prometheus_exporter_utils.cc (hunk 11 above)
Comment (CunjunWang) [reply exporters/prometheus/src/prometheus_exporter_utils.cc]: removed.

**Pull request merged**